test/liquidityConversionRate.js

```diff
@@ -869,6 +879,211 @@ contract('kyberReserve for Liquidity', function(accounts) {
+
+            let rInFp = BigNumber(newRs[id]).mul(formulaPrecision);
+            let pMinInFp = BigNumber(pMin).mul(formulaPrecision);
+            let maxCapBuyInWei = BigNumber(newMaxCapBuyInEth[id]).mul(precision);
+            let maxCapSellInWei = BigNumber(newMaxCapSellInEth[id]).mul(precision);
+            let maxSellRateInPrecision = BigNumber(pMax).mul(precision);
+            let minSellRateInPrecision = BigNumber(pMin).mul(precision);
+
+            await newLiqConvRatesInst.setLiquidityParams(
+                rInFp,
+                pMinInFp,
+                formulaPrecisionBits,
+                maxCapBuyInWei,
+                maxCapSellInWei,
+                feeInBps,
+                maxSellRateInPrecision,
+                minSellRateInPrecision
+            );
+
+            let reserveInst = await Reserve.new(network, newLiqConvRatesInst.address, admin);
+            await reserveInst.setContracts(network, newLiqConvRatesInst.address, 0);
+
+            await newLiqConvRatesInst.setReserveAddress(reserveInst.address);
+
+            //set reserve balance.
+            let reserveEtherInit = (BigNumber(10).pow(18)).mul(newE0s[id]);
+            await Helper.sendEtherWithPromise(accounts[9], reserveInst.address, reserveEtherInit);
+
+            await reserveInst.approveWithdrawAddress(token.address, accounts[0], true);
+
+            //transfer tokens to reserve.
+            let amount = (BigNumber(10).pow(18)).mul(newT0s[id]);
+            await token.transfer(reserveInst.address, amount.valueOf());
+
+            if (i < newRs.length) {
+                // buy then check sell rate
+                let lastBuyRate = BigNumber(precision).div(newP0s[id]).floor();
+                let srcAmount = maxCapBuyInWei;
+                while (true) {
+                    let bal = await token.balanceOf(reserveInst.address);
+                    if (bal * 1.0 <= precision * 1.0) { break; }
+                    let buyRate = await reserveInst.getConversionRate(ethAddress, token.address, srcAmount, currentBlock);
+                    if (buyRate.valueOf() == 0) {
+                        while (true) {
+                            srcAmount = srcAmount.div(2).floor();
+                            if (srcAmount.valueOf() == 0) { break; }
+                            buyRate = await reserveInst.getConversionRate(ethAddress, token.address, srcAmount, currentBlock);
+                            if (buyRate.valueOf() != 0) { break; }
+                        }
+                        if (buyRate.valueOf() == 0) { break; }
+                    }
+                    lastBuyRate = buyRate;
+
+                    await reserveInst.trade(ethAddress, srcAmount, token.address, user1, buyRate, true, {from:network, value: srcAmount});
+                }
+                let minBuyRate = BigNumber(precision).div(pMax).floor();
+                assert(lastBuyRate * 1.0 >= minBuyRate * 1.0, "rate must be greater than min rate, loop: " + i);
```

Comment:
please log last buy rate with min buy rate, to know this test part completed successfully.

must be greater?
not greater equal?

Reply:
please add one check here that shows we get rate 0 when calculated rate is lower then min rate or bigger then max rate.

Reply:
Added checks